### ISSUE TITLE
StopForumSpam confidence levels

### ISSUE THREAD

**Issue by euantorano** [mybb/mybb#3123]
This is related to #2910 and the PR that resolved it - #3121.

Currently, the Stop Forum Spam spam prevention system has a single `confidence` level that encapsulates emaila ddress, username and IP address confidence from the Stop Forum Spam. In the future, it would be a good idea to instead have 4 confidence settings:

- A confidence level for usernames - if the username exceeds this confidence level, block the registration/post/whatever.
- A confidence level for email addresses - if the email address exceeds this confidence level, block the registration/post/whatever.
- A confidence level for IP addresses - if the IP address exceeds this confidence level, block the registration/post/whatever.
- An overall confidence level which is `username + email + ip address / 3` (or by however many things are actually checked) - this will cath cases where the username isn't 100% confident, nor is the email but when both are together, the system is triggered.

We can leave this to 1.9 or a future release, as it's an enahcnement rather than a bug fix.

**Comment by effone** [comment]
Which summarize to have an extra setting activating which the overall score of the above select 3 will be in effect? That leads another question: Check username, Check email & Check overall active but Check IP is off so the system will perform just overall evaluating `username + email / 2` ?

**Comment by euantorano** [comment]
Yes, so there will be 4 settings for confidence. We might be able to then remove the old settings for `check IP` and such (we could say that if `ip confidence == -1` then don't check IP or something like that). If only two items are checked, then the total should be divided by two, yes (the overall would be an `average confidence` and we'd have to update the description to explain that).

**Comment by theonlypwner** [comment]
The calculation proposed in #2910 and implemented in #3121 makes no sense.

Using the average (arithmetic mean) allows spammers to exploit the calculation by combining one factor with a lower confidence score to get a lower overall score. Having an email with 0.99 confidence combined with a clean IP (confidence does not appear in the response for IP) is treated as 0.99, worse than that same email with an IP used to spam with low confidence, treated as around 0.495.

The calculation should return a higher value when additional confidence scores are present. This can be done by simply taking the maximum, or by combining the confidence scores by multiplying the complement of the values.

For example, if you have two confidence values, 0.1 and 0.9,

* Combination method: 1 - (1-0.1) * (1-0.9) = 0.91
* Maximum method: max(0.1, 0.9) = 0.9
* Average method (bad): (0.1 + 0.9) / 2 = 0.5

If you add another confidence value 0.2, the final value should go up even more:

* Combination method: 1 - (1-0.1) * (1-0.9) * (1-0.2) = 0.928
* Maximum method: max(0.1, 0.9, 0.2) = 0.9
* Average method (bad): (0.1 + 0.9 + 0.2) / 3 = 0.4

There might be a better way to combine Wilson scores, but using the average is plainly wrong.
